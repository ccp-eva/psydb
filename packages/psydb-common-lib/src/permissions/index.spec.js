'use strict';
var Permissions = require('./');

describe('Permissions()', () => {
    test('non-root; one research group; single flag', () => {
        var permissions = Permissions({
            hasRootAccess: false,
            researchGroupIds: [ 'gid_1' ],
            forcedResearchGroupId: undefined,
            rolesByResearchGroupId: {
                'gid_1': { state: {
                    canReadSubjects: true,
                }}
            }
        });

        //console.dir({ permissions }, { depth: null })
        expect(permissions).toMatchSnapshot()
    });

    test('non-root; multiple research group; single flag', () => {
        var permissions = Permissions({
            hasRootAccess: false,
            researchGroupIds: [ 'gid_1', 'gid_2' ],
            forcedResearchGroupId: undefined,
            rolesByResearchGroupId: {
                'gid_1': { state: {
                    canReadSubjects: true,
                }},
                'gid_2': { state: {
                    canReadSubjects: true,
                }}
            }
        });

        //console.dir({ permissions }, { depth: null })
        expect(permissions).toMatchSnapshot()
    })
    
    test('non-root; multiple research group; mixed flags', () => {
        var permissions = Permissions({
            hasRootAccess: false,
            researchGroupIds: [ 'gid_1', 'gid_2' ],
            forcedResearchGroupId: undefined,
            rolesByResearchGroupId: {
                'gid_1': { state: {
                    canReadSubjects: true,
                    canWriteSubjects: true,
                }},
                'gid_2': { state: {
                    canReadSubjects: true,
                    canReadStudies: true,
                }}
            }
        });

        //console.dir({ permissions }, { depth: null })
        expect(permissions).toMatchSnapshot()
    })

    test('non-root; multiple research group + forced; mixed flags', () => {
        var permissions = Permissions({
            hasRootAccess: false,
            researchGroupIds: [ 'gid_1', 'gid_2' ],
            forcedResearchGroupId: 'gid_2',
            rolesByResearchGroupId: {
                'gid_1': { state: {
                    canReadSubjects: true,
                    canWriteSubjects: true,
                }},
                'gid_2': { state: {
                    canReadSubjects: true,
                    canReadStudies: true,
                }}
            }
        });

        //console.dir({ permissions }, { depth: null })
        expect(permissions).toMatchSnapshot()
    })

    test('root; multiple research group + forced; mixed flags', () => {
        var permissions = Permissions({
            hasRootAccess: true,
            researchGroupIds: [ 'gid_1', 'gid_2' ],
            forcedResearchGroupId: 'gid_2',
            rolesByResearchGroupId: {
                'gid_1': { state: {
                    canReadSubjects: true,
                    canWriteSubjects: true,
                }},
                'gid_2': { state: {
                    canReadSubjects: true,
                    canReadStudies: true,
                }}
            }
        });

        //console.dir({ permissions }, { depth: null })
        expect(permissions).toMatchSnapshot()
    })
})
