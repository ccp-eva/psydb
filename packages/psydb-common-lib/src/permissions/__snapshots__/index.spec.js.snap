// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permissions() non-root; multiple research group + forced; mixed flags 1`] = `
Object {
  "availableResearchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "flagsByResearchGroupId": Object {
    "gid_1": Object {
      "canReadSubjects": true,
      "canWriteSubjects": true,
    },
    "gid_2": Object {
      "canReadStudies": true,
      "canReadSubjects": true,
    },
  },
  "forcedResearchGroupId": "gid_2",
  "hasRootAccess": false,
  "researchGroupIds": Array [
    "gid_2",
  ],
  "researchGroupIdsByCollection": Object {
    "ageFrame": Object {
      "read": Array [],
      "write": Array [],
    },
    "customRecordType": Object {
      "read": Array [],
      "write": Array [],
    },
    "experiment": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentOperatorTeam": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariant": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariantSetting": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalOrganization": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalPerson": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSet": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSetItem": Object {
      "read": Array [],
      "write": Array [],
    },
    "location": Object {
      "read": Array [],
      "write": Array [],
    },
    "personnel": Object {
      "read": Array [],
      "write": Array [],
    },
    "researchGroup": Object {
      "read": Array [],
      "write": Array [],
    },
    "reservation": Object {
      "read": Array [],
      "write": Array [],
    },
    "study": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subject": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subjectSelector": Object {
      "read": Array [],
      "write": Array [],
    },
    "systemRole": Object {
      "read": Array [],
      "write": Array [],
    },
  },
  "researchGroupIdsByFlag": Object {
    "canReadStudies": Array [
      "gid_2",
    ],
    "canReadSubjects": Array [
      "gid_2",
    ],
  },
}
`;

exports[`Permissions() non-root; multiple research group; deep flags 1`] = `
Object {
  "availableResearchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "flagsByResearchGroupId": Object {
    "gid_1": Object {
      "canReadSubjects": true,
      "canWriteSubjects": true,
      "labOperations": Object {
        "inhouse": Object {
          "canMoveAndCancelExperiments": true,
          "canViewExperimentCalendar": true,
        },
      },
    },
    "gid_2": Object {
      "canReadStudies": true,
      "canReadSubjects": true,
      "labOperations": Object {
        "inhouse": Object {
          "canMoveAndCancelExperiments": false,
          "canViewExperimentCalendar": true,
        },
      },
    },
  },
  "forcedResearchGroupId": undefined,
  "hasRootAccess": false,
  "researchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "researchGroupIdsByCollection": Object {
    "ageFrame": Object {
      "read": Array [],
      "write": Array [],
    },
    "customRecordType": Object {
      "read": Array [],
      "write": Array [],
    },
    "experiment": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentOperatorTeam": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariant": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariantSetting": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalOrganization": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalPerson": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSet": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSetItem": Object {
      "read": Array [],
      "write": Array [],
    },
    "location": Object {
      "read": Array [],
      "write": Array [],
    },
    "personnel": Object {
      "read": Array [],
      "write": Array [],
    },
    "researchGroup": Object {
      "read": Array [],
      "write": Array [],
    },
    "reservation": Object {
      "read": Array [],
      "write": Array [],
    },
    "study": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subject": Object {
      "read": Array [
        "gid_1",
        "gid_2",
      ],
      "write": Array [
        "gid_1",
      ],
    },
    "subjectSelector": Object {
      "read": Array [],
      "write": Array [],
    },
    "systemRole": Object {
      "read": Array [],
      "write": Array [],
    },
  },
  "researchGroupIdsByFlag": Object {
    "canReadStudies": Array [
      "gid_2",
    ],
    "canReadSubjects": Array [
      "gid_1",
      "gid_2",
    ],
    "canWriteSubjects": Array [
      "gid_1",
    ],
    "labOperations": Object {
      "inhouse": Object {
        "canMoveAndCancelExperiments": Array [
          "gid_1",
        ],
        "canViewExperimentCalendar": Array [
          "gid_1",
          "gid_2",
        ],
      },
    },
  },
}
`;

exports[`Permissions() non-root; multiple research group; mixed flags 1`] = `
Object {
  "availableResearchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "flagsByResearchGroupId": Object {
    "gid_1": Object {
      "canReadSubjects": true,
      "canWriteSubjects": true,
    },
    "gid_2": Object {
      "canReadStudies": true,
      "canReadSubjects": true,
    },
  },
  "forcedResearchGroupId": undefined,
  "hasRootAccess": false,
  "researchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "researchGroupIdsByCollection": Object {
    "ageFrame": Object {
      "read": Array [],
      "write": Array [],
    },
    "customRecordType": Object {
      "read": Array [],
      "write": Array [],
    },
    "experiment": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentOperatorTeam": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariant": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariantSetting": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalOrganization": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalPerson": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSet": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSetItem": Object {
      "read": Array [],
      "write": Array [],
    },
    "location": Object {
      "read": Array [],
      "write": Array [],
    },
    "personnel": Object {
      "read": Array [],
      "write": Array [],
    },
    "researchGroup": Object {
      "read": Array [],
      "write": Array [],
    },
    "reservation": Object {
      "read": Array [],
      "write": Array [],
    },
    "study": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subject": Object {
      "read": Array [
        "gid_1",
        "gid_2",
      ],
      "write": Array [
        "gid_1",
      ],
    },
    "subjectSelector": Object {
      "read": Array [],
      "write": Array [],
    },
    "systemRole": Object {
      "read": Array [],
      "write": Array [],
    },
  },
  "researchGroupIdsByFlag": Object {
    "canReadStudies": Array [
      "gid_2",
    ],
    "canReadSubjects": Array [
      "gid_1",
      "gid_2",
    ],
    "canWriteSubjects": Array [
      "gid_1",
    ],
  },
}
`;

exports[`Permissions() non-root; multiple research group; single flag 1`] = `
Object {
  "availableResearchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "flagsByResearchGroupId": Object {
    "gid_1": Object {
      "canReadSubjects": true,
    },
    "gid_2": Object {
      "canReadSubjects": true,
    },
  },
  "forcedResearchGroupId": undefined,
  "hasRootAccess": false,
  "researchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "researchGroupIdsByCollection": Object {
    "ageFrame": Object {
      "read": Array [],
      "write": Array [],
    },
    "customRecordType": Object {
      "read": Array [],
      "write": Array [],
    },
    "experiment": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentOperatorTeam": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariant": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariantSetting": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalOrganization": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalPerson": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSet": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSetItem": Object {
      "read": Array [],
      "write": Array [],
    },
    "location": Object {
      "read": Array [],
      "write": Array [],
    },
    "personnel": Object {
      "read": Array [],
      "write": Array [],
    },
    "researchGroup": Object {
      "read": Array [],
      "write": Array [],
    },
    "reservation": Object {
      "read": Array [],
      "write": Array [],
    },
    "study": Object {
      "read": Array [],
      "write": Array [],
    },
    "subject": Object {
      "read": Array [
        "gid_1",
        "gid_2",
      ],
      "write": Array [],
    },
    "subjectSelector": Object {
      "read": Array [],
      "write": Array [],
    },
    "systemRole": Object {
      "read": Array [],
      "write": Array [],
    },
  },
  "researchGroupIdsByFlag": Object {
    "canReadSubjects": Array [
      "gid_1",
      "gid_2",
    ],
  },
}
`;

exports[`Permissions() non-root; one research group; single flag 1`] = `
Object {
  "availableResearchGroupIds": Array [
    "gid_1",
  ],
  "flagsByResearchGroupId": Object {
    "gid_1": Object {
      "canReadSubjects": true,
    },
  },
  "forcedResearchGroupId": undefined,
  "hasRootAccess": false,
  "researchGroupIds": Array [
    "gid_1",
  ],
  "researchGroupIdsByCollection": Object {
    "ageFrame": Object {
      "read": Array [],
      "write": Array [],
    },
    "customRecordType": Object {
      "read": Array [],
      "write": Array [],
    },
    "experiment": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentOperatorTeam": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariant": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariantSetting": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalOrganization": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalPerson": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSet": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSetItem": Object {
      "read": Array [],
      "write": Array [],
    },
    "location": Object {
      "read": Array [],
      "write": Array [],
    },
    "personnel": Object {
      "read": Array [],
      "write": Array [],
    },
    "researchGroup": Object {
      "read": Array [],
      "write": Array [],
    },
    "reservation": Object {
      "read": Array [],
      "write": Array [],
    },
    "study": Object {
      "read": Array [],
      "write": Array [],
    },
    "subject": Object {
      "read": Array [
        "gid_1",
      ],
      "write": Array [],
    },
    "subjectSelector": Object {
      "read": Array [],
      "write": Array [],
    },
    "systemRole": Object {
      "read": Array [],
      "write": Array [],
    },
  },
  "researchGroupIdsByFlag": Object {
    "canReadSubjects": Array [
      "gid_1",
    ],
  },
}
`;

exports[`Permissions() root; multiple research group + forced; deep flags 1`] = `
Object {
  "availableResearchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "flagsByResearchGroupId": Object {
    "gid_1": Object {
      "canReadSubjects": true,
      "canWriteSubjects": true,
      "labOperations": Object {
        "inhouse": Object {
          "canMoveAndCancelExperiments": true,
          "canViewExperimentCalendar": true,
        },
      },
    },
    "gid_2": Object {
      "canReadStudies": true,
      "canReadSubjects": true,
      "labOperations": Object {
        "inhouse": Object {
          "canMoveAndCancelExperiments": false,
          "canViewExperimentCalendar": true,
        },
      },
    },
  },
  "forcedResearchGroupId": "gid_2",
  "hasRootAccess": true,
  "researchGroupIds": Array [
    "gid_2",
  ],
  "researchGroupIdsByCollection": Object {
    "ageFrame": Object {
      "read": Array [],
      "write": Array [],
    },
    "customRecordType": Object {
      "read": Array [],
      "write": Array [],
    },
    "experiment": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentOperatorTeam": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariant": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariantSetting": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalOrganization": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalPerson": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSet": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSetItem": Object {
      "read": Array [],
      "write": Array [],
    },
    "location": Object {
      "read": Array [],
      "write": Array [],
    },
    "personnel": Object {
      "read": Array [],
      "write": Array [],
    },
    "researchGroup": Object {
      "read": Array [],
      "write": Array [],
    },
    "reservation": Object {
      "read": Array [],
      "write": Array [],
    },
    "study": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subject": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subjectSelector": Object {
      "read": Array [],
      "write": Array [],
    },
    "systemRole": Object {
      "read": Array [],
      "write": Array [],
    },
  },
  "researchGroupIdsByFlag": Object {
    "canReadStudies": Array [
      "gid_2",
    ],
    "canReadSubjects": Array [
      "gid_2",
    ],
    "labOperations": Object {
      "inhouse": Object {
        "canViewExperimentCalendar": Array [
          "gid_2",
        ],
      },
    },
  },
}
`;

exports[`Permissions() root; multiple research group + forced; mixed flags 1`] = `
Object {
  "availableResearchGroupIds": Array [
    "gid_1",
    "gid_2",
  ],
  "flagsByResearchGroupId": Object {
    "gid_1": Object {
      "canReadSubjects": true,
      "canWriteSubjects": true,
    },
    "gid_2": Object {
      "canReadStudies": true,
      "canReadSubjects": true,
    },
  },
  "forcedResearchGroupId": "gid_2",
  "hasRootAccess": true,
  "researchGroupIds": Array [
    "gid_2",
  ],
  "researchGroupIdsByCollection": Object {
    "ageFrame": Object {
      "read": Array [],
      "write": Array [],
    },
    "customRecordType": Object {
      "read": Array [],
      "write": Array [],
    },
    "experiment": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentOperatorTeam": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariant": Object {
      "read": Array [],
      "write": Array [],
    },
    "experimentVariantSetting": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalOrganization": Object {
      "read": Array [],
      "write": Array [],
    },
    "externalPerson": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSet": Object {
      "read": Array [],
      "write": Array [],
    },
    "helperSetItem": Object {
      "read": Array [],
      "write": Array [],
    },
    "location": Object {
      "read": Array [],
      "write": Array [],
    },
    "personnel": Object {
      "read": Array [],
      "write": Array [],
    },
    "researchGroup": Object {
      "read": Array [],
      "write": Array [],
    },
    "reservation": Object {
      "read": Array [],
      "write": Array [],
    },
    "study": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subject": Object {
      "read": Array [
        "gid_2",
      ],
      "write": Array [],
    },
    "subjectSelector": Object {
      "read": Array [],
      "write": Array [],
    },
    "systemRole": Object {
      "read": Array [],
      "write": Array [],
    },
  },
  "researchGroupIdsByFlag": Object {
    "canReadStudies": Array [
      "gid_2",
    ],
    "canReadSubjects": Array [
      "gid_2",
    ],
  },
}
`;
